add releasing.md (#321)
metal3-io/metal3-docs: processes/releasing.md

# Releasing

This document details the high-level process of creating a Metal3 release, and
the post-release actions required to add new release branches to the CI.

Note the [TODO](#todo) section in the end.

## Process

High-level release process we followed in case it makes it clearer:

1. [Release IPAM](https://github.com/metal3-io/ip-address-manager/blob/main/docs/releasing.md)
1. [Release BMO](https://github.com/metal3-io/baremetal-operator/blob/main/docs/releasing.md)
1. [Release CAPM3](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/docs/releasing.md)
1. Update Metal3 dev-env and CI to support new releases

## Update Metal3 dev-env and CI to support new releases

> If a new minor or major was released, we need to make changes in Metal3 dev-env

Reviewer: ```suggestion
If a new minor or major release was published, we need to make changes in Metal3 dev-env
```

and CI.

### Tag Ironic and Mariadb images

After releasing `v1.x.y` version of CAPM3, create an annotated tag with `capm3-`
prefix + release version in
[ironic-image](https://github.com/metal3-io/ironic-image) and
[mariadb-image](https://github.com/metal3-io/mariadb-image) Github
repositories.

```bash
git tag -s -a capm3-v1.x.y -m capm3-v1.x.y [optional sha, or tag^{}]
git push origin capm3-v1.x.y
```

After this, [Ironic](https://quay.io/repository/metal3-io/ironic) and
[MariaDB](https://quay.io/repository/metal3-io/mariadb) container images will
be built automatically  with `capm3-v1.x.y` tag in Quay. Verify the build is
triggered, as often Quay has disabled the build trigger due build failures or
other reasons.

### Create Jenkins jobs

Two Jenkins jobs need to be created in
[JJB](https://gerrit.nordix.org/plugins/gitiles/infra/cicd/+/refs/heads/master/jjb/metal3/):

- a periodic job that runs on a regular basis.
- a PR verification job that is triggered by a keyword on a PR targeted for that
  release branch.

> [Gerrit PR](https://gerrit.nordix.org/c/infra/cicd/+/17709) that adds support

Reviewer: Something like, "Prior art" (commonly used in CAPI for referencing example previous PR) may suit here and in all occurrences below (Example PR) better:
```suggestion
[Prior art](https://gerrit.nordix.org/c/infra/cicd/+/17709) that adds support
```

Author: Fixed

for periodics and PR jobs.

### Change project-infra to support new branches

We also need to change project-infra keywords for required tests for some of
the repositories.

[Example PR](https://github.com/metal3-io/project-infra/pull/496)

> 

Reviewer: Once project infra PR is merged (assuming CI jobs are ready, release branches exits and so on), the new jobs that are made as required in the infra PR (i.e in above PR, it is `test-ubuntu-integration-release-1-4`) has to be also configured in corresponding GitHub repo branch protection rules, otherwise it is useless and GitHub won't be seeing them as mandatory. Let's take a CAPM3 as an example and how to do it:

- trigger a new job (`test-ubuntu-integration-release-1-4`) in the PR (could be test PR) pointing to that branch (release-1.4) at least once so that it can be configured in later steps
- someone with admin rights in the org, has to go to https://github.com/metal3-io/cluster-api-provider-metal3/settings/branches > Click `Edit` on release-1.4 branch > Paste the new test job in search bar in `Status checks that are required.` and select it from drop down > Click on "Save changes"

Author: This is documented per repository in their release documentation, see linked PRs.

> ### Configure dev-env to support new branches

Reviewer: for consistency:
```suggestion
### Configure Metal3 dev-env to support new branches
```

Author: Fixed

Metal3-dev-env needs to be modified to support new release branches and
tags with supported combination of component versions for testing.

[Example PR](https://github.com/metal3-io/metal3-dev-env/pull/1222)

### Configure dev-tools to sync new release branches to Nordix

> Dev-tools syncronizes `metal3-io` changes to the `Nordix` fork for the EST

Reviewer: ```suggestion
Dev-tools synchronises `metal3-io` changes to the `Nordix` fork for the EST
```

Author: Fixed

git push origin capm3-v1.x.y
```

After this, [Ironic](https://quay.io/repository/metal3-io/ironic) and
[MariaDB](https://quay.io/repository/metal3-io/mariadb) container images will
be built automatically  with `capm3-v1.x.y` tag in Quay. Verify the build is
triggered, as often Quay has disabled the build trigger due build failures or
other reasons.

### Create Jenkins jobs

Two Jenkins jobs need to be created in
[JJB](https://gerrit.nordix.org/plugins/gitiles/infra/cicd/+/refs/heads/master/jjb/metal3/):

- a periodic job that runs on a regular basis.
- a PR verification job that is triggered by a keyword on a PR targeted for that
  release branch.

[Gerrit PR](https://gerrit.nordix.org/c/infra/cicd/+/17709) that adds support
for periodics and PR jobs.

### Change project-infra to support new branches

We also need to change project-infra keywords for required tests for some of
the repositories.

[Project infra PR](https://github.com/metal3-io/project-infra/pull/496)

### Configure dev-env to support new branches

Metal3-dev-env variables need to be modified. After a major or minor release,
the new minor version should point to main for CAPM3/IPAM and the released

> version should point to the release branch.

Reviewer 2: I'm not understanding what this means. Is it something about the combinations that should be tested?

Author: Yes, and the dev-env example PR is there to help understand what needs to be changed. I can try reword it a bit.

Author: Done, simplified it. Is it any better now?

Reviewer 2: Thanks much clearer!

**WARNING**: this breaks metal3-dev-env (CI included) until dev-env and
IPAM/CAPM3 repositories are in sync.

> [Dev-env PR](https://github.com/metal3-io/metal3-dev-env/pull/1222) TODO

Reviewer 2: ```suggestion
[Dev-env PR](https://github.com/metal3-io/metal3-dev-env/pull/1222)
```

Author: Whoops, fixed.

# Releasing

WIP: This document is work in progress.

## Process

High-level release process we followed in case it makes it clearer:

1. Uplift IPAM to use latest CAPI
1. [Release IPAM](https://github.com/metal3-io/ip-address-manager/blob/main/docs/releasing.md)
1. Uplift BMO to use latest go modules as in CAPI (controller runtime,
   `k8s.io/*` in `go.mod`)
1. [Release BMO](https://github.com/metal3-io/baremetal-operator/blob/main/docs/releasing.md)
1. Uplift CAPM3 to use latest CAPI, IPAM, BMO (in `go.mod`)
1. [Release CAPM3](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/docs/releasing.md)
1. Update Metal3 dev-env and CI to support new releases

## Update Metal3 dev-env and CI to support new releases

If a new minor or major was released, we need to make changes in Metal3 dev-env
and CI.

### Tag Ironic and Mariadb images

After releasing v1.x.y version of CAPM3, create an annotated tag with `capm3-`
prefix + release version in

> [Ironic-image](https://github.com/metal3-io/ironic-image) and

Reviewer 3: nit: ironic-image

Author: Fixed

[mariadb-image](https://github.com/metal3-io/mariadb-image) Github
repositories.

```bash
git tag -s -a capm3-v1.x.y -m capm3-v1.x.y [optional sha, or tag^{}]
git push origin capm3-v1.x.y
```

After this, [Ironic](https://quay.io/repository/metal3-io/ironic) and

> [Mariadb](https://quay.io/repository/metal3-io/mariadb) container images will

Reviewer 3: nit: MariaDB

Author: Fixed

be built automatically  with `capm3-v1.x.y` tag in Quay. Verify the build is
triggered, as often Quay has disabled the build trigger due build failures or
other reasons.

### Create Jenkins jobs

Two Jenkins jobs need to be created in
[JJB](https://gerrit.nordix.org/plugins/gitiles/infra/cicd/+/refs/heads/master/jjb/metal3/):

- a periodic job that runs on a regular basis.
- a PR verification job that is triggered by a keyword on a PR targeted for that
  release branch.

We also need to change project-infra keywords for required tests for some of
the repositories.

### Dev-env changes

> Metal3-dev-env variables need to be modified. After a major or minor release,

Reviewer 3: Maybe provide an example or at least the line(s) to change?

Author: That's the plan. This PR is marked WIP as I'm documenting the required changes as were doing them in 1.4 release cycle.

Author: Done.

the new minor version should point to main for CAPM3/IPAM and the released
version should point to the release branch.

WARNING: this breaks metal3-dev-env (CI included) until dev-env and IPAM/CAPM3
repositories are in sync.

### Update testing matrix documentation

Update [testing matrix documentation](https://github.com/metal3-io/metal3-docs/pull/320)
to reflect the new release and the combination of components that are tested in
periodic jobs.

TODO: change PR link to document link after PR is merged.

## BMO releasing to be discussed

1. BMO releasing and maintenance
   - BMO does not have release branches, only tags off the `main` branch
   - Without release branches, maintaining BMO versions that are coupled with CAPM3 releases is impossible
1. BMO contains multiple subcomponents and images:
   - keepalived ([within BMO repo](https://github.com/metal3-io/baremetal-operator/tree/main/resources/keepalived-docker))
   - [mariadb-image](https://github.com/metal3-io/mariadb-image)

>    - [ironic-image](https://github.com/metal3-io/ironic-image)

Reviewer 3: BMO does not contain these two, maybe change the wording?

Author: This is WIP. The wording will be changed quite a bit, especially everything around BMO.

Author: Done. Let me know if the wording is OK, and if not, please propose a better wording.

Reviewer 3: I'm not sure what you're intending to say. Maybe "BMO makes strong assumption about the exact configuration of Ironic. However, [ironic-image](https://github.com/metal3-io/ironic-image) is currently branchless."

Author: Thanks. Used your wording and cut down a lot of the details, they were unnecessary here.
